optel/optel.go

```diff
@@ -0,0 +1,197 @@
+	}
+
+	sd = func() {
+		err = errors.Join(err, shutdown(context.Background()))
+	}
+
+	// handleErr calls shutdown for cleanup and makes sure that all errors are returned.
+	handleErr := func(inErr error) {
+		err = errors.Join(inErr, shutdown(ctx))
+	}
+
+	globalResource, err = resource.Merge(
+		resource.Default(),
+		resource.NewWithAttributes(
+			semconv.SchemaURL,
+			semconv.ServiceName(config.Appname),
+		),
+	)
+	if err != nil {
+		handleErr(err)
+		return
+	}
+
+	// Set up propagator.
+	prop := newPropagator()
+	otel.SetTextMapPropagator(prop)
+
+	// Set up trace provider.
+	tracerProvider, err := newTraceProvider(ctx)
+	if err != nil {
+		handleErr(err)
+		return
+	}
+
+	shutdownFuncs = append(shutdownFuncs, tracerProvider.Shutdown)
+	otel.SetTracerProvider(tracerProvider)
+	globalTracer = otel.Tracer("")
+	fmt.Println("Tracer provider set")
+
+	return
+}
+
+func newPropagator() propagation.TextMapPropagator {
+	return propagation.NewCompositeTextMapPropagator(
+		propagation.TraceContext{},
+	)
+}
+
+func newStdoutTraceProvider(ctx context.Context) (*trace.TracerProvider, error) {
+	traceExporter, err := stdouttrace.New(
+		stdouttrace.WithPrettyPrint())
+	if err != nil {
+		return nil, err
+	}
+	traceProvider := trace.NewTracerProvider(
+		trace.WithResource(globalResource),
+		trace.WithBatcher(traceExporter),
+	)
+	return traceProvider, nil
+}
+
+func newTraceProvider(ctx context.Context) (*trace.TracerProvider, error) {
+	traceExporter, err := otlptracehttp.New(ctx, otlptracehttp.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+	traceProvider := trace.NewTracerProvider(
+		trace.WithResource(globalResource),
+		trace.WithBatcher(traceExporter),
+	)
+	return traceProvider, nil
+}
+
+func getReqIdMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		defer StartTrack(ctx, "ReqID")()
+		next.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+func getFilteredOtelchi(appName string, r chi.Routes) func(next http.Handler) http.Handler {
+	healthFilter := func(r *http.Request) bool {
+		if r.URL.Path == "/health" {
+			return false
+		}
+		return true
+	}
+	return otelchi.Middleware(
+		appName, otelchi.WithChiRoutes(r),
+		otelchi.WithFilter(healthFilter),
+	)
+}
+
+func TraceMiddlewares(appName string, r chi.Routes) []func(next http.Handler) http.Handler {
+	var middlewares []func(next http.Handler) http.Handler
+	middlewares = append(middlewares, getFilteredOtelchi(appName, r))
+	middlewares = append(middlewares, getReqIdMiddleware)
+	return middlewares
+}
+
+func addCTXTraceAttributes(ctx context.Context, s *oteltrace.Span) {
+	// Add Request ID if found
+	(*s).SetAttributes(attribute.String(requestIDField, dmMiddleware.GetReqID(ctx)))
+
+	// Add configured CTXAttributes if found
+	for k, v := range config.TraceConfig.CTXAttributes {
+		if _v, ok := ctx.Value(k).(string); ok {
+			(*s).SetAttributes(attribute.String(v, _v))
+		}
+	}
+}
```

Comment:
Dois pontos aqui:

- O `oteltrace.Span` é uma interface, não uma struct, você não precisa (e não deve) tratar ela como um ponteiro.
Por padrão de implementação no Go, a struct recebida por trás dessa interface será um ponteiro 😉 

- É bom dar uma melhorada na nomenclatura das variáveis para facilitar um pouco a leitura do código.
Na DM a gente abandonou um pouco essa prática de instanciar os valores de variáveis com letras hehe

Tenta fazer isso e ve se funciona bem:
```suggestion
func addCTXTraceAttributes(ctx context.Context, s oteltrace.Span) {
	// Add Request ID if found
	s.SetAttributes(attribute.String(requestIDField, dmMiddleware.GetReqID(ctx)))

	// Add configured CTXAttributes if found
	for contextKey, key := range config.TraceConfig.CTXAttributes {
		if value, ok := ctx.Value(contextKey).(string); ok {
			s.SetAttributes(attribute.String(key, value))
		}
	}
}
```